src/com/esotericsoftware/yamlbeans/YamlReader.java

```diff
@@ -1,424 +1,435 @@
+				} else if (type == Character.class) {
+					convertedValue = value.length() == 0 ? null : value.charAt(0);
+				} else if (type == Byte.TYPE) {
+					convertedValue = value.length() == 0 ? 0 : Byte.decode(value);
+				} else if (type == Byte.class) {
+					convertedValue = value.length() == 0 ? null : Byte.decode(value);
+				} else
+					throw new YamlException("Unknown field type.");
+				if (anchor != null) anchors.put(anchor, convertedValue);
+				return convertedValue;
+			} catch (Exception ex) {
+				throw new YamlReaderException("Unable to convert value to required type \"" + type + "\": " + value, ex);
+			}
+		}
+
+		if (Enum.class.isAssignableFrom(type)) {
+			Event event = parser.getNextEvent();
+			if (event.type != SCALAR) throw new YamlReaderException("Expected scalar for enum type but found: " + event.type);
+			String enumValueName = ((ScalarEvent)event).value;
+			if (enumValueName.length() == 0) return null;
+			try {
+				return Enum.valueOf(type, enumValueName);
+			} catch (Exception ex) {
+				throw new YamlReaderException("Unable to find enum value '" + enumValueName + "' for enum class: " + type.getName());
+			}
+		}
+
+		for (Entry<Class, ScalarSerializer> entry : config.scalarSerializers.entrySet()) {
+			if (entry.getKey().isAssignableFrom(type)) {
+				ScalarSerializer serializer = entry.getValue();
+				Event event = parser.getNextEvent();
+				if (event.type != SCALAR)
+					throw new YamlReaderException("Expected scalar for type '" + type + "' to be deserialized by scalar serializer '"
+						+ serializer.getClass().getName() + "' but found: " + event.type);
+				Object value = serializer.read(((ScalarEvent)event).value);
+				if (anchor != null) anchors.put(anchor, value);
+				return value;
+			}
+		}
+
+		Event event = parser.peekNextEvent();
+		switch (event.type) {
+		case MAPPING_START: {
+			// Must be a map or an object.
+			event = parser.getNextEvent();
+			Object object;
+			try {
+				object = createObject(type);
+			} catch (InvocationTargetException ex) {
+				throw new YamlReaderException("Error creating object.", ex);
+			}
+			if (anchor != null) anchors.put(anchor, object);
+			ArrayList keys = new ArrayList();
+			while (true) {
+				if (parser.peekNextEvent().type == MAPPING_END) {
+					parser.getNextEvent();
+					break;
+				}
+				Object key = readValue(null, null, null);
+				// Explicit key/value pairs (using "? key\n: value\n") will come back as a map.
+				boolean isExplicitKey = key instanceof Map;
+				Object value = null;
+				if (isExplicitKey) {
+					Entry nameValuePair = (Entry)((Map)key).entrySet().iterator().next();
+					key = nameValuePair.getKey();
+					value = nameValuePair.getValue();
+				}
+				if (object instanceof Map) {
+					// Add to map.
+					if (!isExplicitKey) value = readValue(elementType, null, null);
+					if (!allowDuplicates && keys.contains(key)) {
+						throw new YamlReaderException("Duplicate key found '" + key + "'");
+					}
+					keys.add(key);
```

Comment:
`object` is a Map, just check if the map already has the value. Don't create an ArrayList when `object` is a Map. When it is NOT a map, then you need to use the `keys` ArrayList. Currently you don't find duplicate keys at all when not a map.